packages/filesystem/src/browser/location/location-renderer.tsx

```diff
@@ -18,32 +18,191 @@ import URI from '@theia/core/lib/common/uri';
 import { LocationService } from './location-service';
 import { ReactRenderer } from '@theia/core/lib/browser/widgets/react-renderer';
 import * as React from 'react';
+import * as ReactDOM from 'react-dom';
+import { FileService } from '../file-service';
+import { DisposableCollection, Emitter } from '@theia/core/lib/common';
 
+interface AutoSuggestDataEvent {
+    input: string;
+    parent: string;
+    children: string[];
+}
+
+class ResolvedDirectoryCache {
+    protected pendingResolvedDirectories = new Map<string, { input: string, pendingDirectories: Promise<void> }>();
+    protected cachedDirectories = new Map<string, string[]>();
+
+    protected directoryResolvedEmitter = new Emitter<AutoSuggestDataEvent>();
+    readonly onDirectoryDidResolve = this.directoryResolvedEmitter.event;
+
+    constructor(protected readonly fileService: FileService) { }
+
+    tryResolveChildDirectories(inputAsURI: URI): string[] | undefined {
+        const parentDirectory = inputAsURI.path.dir.toString();
+        const input = inputAsURI.path.toString();
+        if (this.cachedDirectories.has(parentDirectory)) {
+            return this.cachedDirectories.get(parentDirectory);
+        } else if (!this.pendingResolvedDirectories.has(parentDirectory)) {
+            const pendingDirectories = this.createResolutionPromise(parentDirectory);
+            this.pendingResolvedDirectories.set(parentDirectory, { input, pendingDirectories: pendingDirectories });
+        } else if (this.pendingResolvedDirectories.has(parentDirectory)) {
+            // if promise has still not resolved, update its input value to latest
+            const pendingDirectoriesData = (this.pendingResolvedDirectories.get(parentDirectory));
+            if (pendingDirectoriesData) {
+                pendingDirectoriesData.input = input;
+            }
+        }
+        return undefined;
+    }
+
+    protected async createResolutionPromise(directoryToResolve: string): Promise<void> {
+        return this.fileService.resolve(new URI(directoryToResolve)).then(({ children }) => {
+            if (children) {
+                const latestInput = this.pendingResolvedDirectories.get(directoryToResolve)?.input || '';
+                const childDirectories = children.filter(child => child.isDirectory)
+                    .map(directory => `${directory.resource.path}/`)
+                    .sort();
```

Comment:
Do they come out of (alphabetical) order to begin with? Usually `fs.readDir` returns things in alphabetical order, but perhaps `resolve` behaves differently.

Reply:
You're right, I've removed `.sort()`